Migrate coreclr's worker thread pool to be able to use the portable thread pool in opt-in fashion

kouvel opened the pull request:
- Enables using the portable thread pool with coreclr as opt-in. Change is off by default for now, and can be enabled with COMPlus_ThreadPool_UsePortableThreadPool=1. Once it's had bake time and seen to be stable, at a reasonable time in the future the config flag would ideally be removed and the relevant parts of native implementation deleted.
- The IO thread pool is not being migrated in this change, and remains on the native side
- My goal was to get compatible behavior, compatible with diagnostics tools, and similar perf to the native implementation in coreclr. Tried to avoid changing scheduling behavior, behavior of heuristics, etc., compared with that implementation.
- The eventual goal is to have one mostly managed thread pool implementation that can be shared between runtimes, to ease maintenance going forward

Commit descriptions:
- "Add dependencies"
  - Ported LowLevelLock from CoreRT, and moved LowLevelSpinWaiter to shared. Since we support Thread.Interrupt(), they were necessary in the wait subsystem in CoreRT partly to support that, and were also used in the portable thread pool implementation where a pending thread interrupt on a thread pool thread would otherwise crash the process. Interruptible waits are already used in the managed side of the thread pool in the queue implementations. It may be reasonable to ignore the thread interrupt problem and suggest that it not be used on thread pool threads, but for now I just brought in the dependencies to keep behavior consistent with the native implementation.
- "Add config var"
  - Added config var COMPlus_ThreadPool_UsePortableThreadPool (disabled by default for now)
  - Flowed the new config var to the managed side and set up a mechanism to flow all of the thread pool config vars
  - Removed debug-only config var COMPlus_ThreadpoolTickCountAdjustment, which didn't seem to be too useful
  - Specialized native and managed thread pool paths based on the config var. Added assertions to paths that should not be reached depending on the config var.
- "Move portable RegisteredWaitHandle implementation to shared ThreadPool.cs"
  - Just moved the portable implementation, no functional changes. In preparation for merging the two implementations.
- "Merge RegisteredWaitHandle implementations"
  - Merged implementations of RegisteredWaitHandle using the portable version as the primary and specializing small parts of it for coreclr
  - Fixed PortableThreadPool's registered waits to track SafeWaitHandles instead of WaitHandles similarly to the native implementation. The SafeWaitHandle in a WaitHandle can be modified, so it is retrieved once and reused thereafter. Also added/removed refs for the SafeWaitHandles that are registered.
- "Separate portable-only portion of RegisteredWaitHandle"
  - Separated RegisteredWaitHandle.UnregisterPortable into a different file, no functional changes. Those paths reference PortableThreadPool, which is conditionally included unlike ThreadPool.cs. Just for consistency such that the new file can be conditionally included similarly to PortableThreadPool.
- "Fix timers, tiered compilation, introduced time-sensitive work item queue to simulate coreclr behavior"
  - Wired work items queued from the native side (appdomain timer callback, tiered compilation background work callback) to queue them into the managed side
  - The timer thread calls into managed code to queue the callback
  - Some tiered compilation work item queuing paths cannot call managed code, so used a timer with zero due time instead
  - Added a queue of "time-sensitive" work items to the managed side to mimic how work items queued from the native side ran previously. In particular, if the global queue is backed up, when using the native thread pool the native work items still run ahead of them periodically (based on the Dispatch quantum). Potentially they could be queued into the global queue but if it's backed up it can potentially significantly and perhaps artificially delay the appdomain timer callback and the tiering background jitting. I didn't want to change the behavior in an observable (and potentially bad) way here for now, a good time to revisit this would be when IO completion handling is added to the portable thread pool, then the native work items could be handled somewhat similarly.
- "Implement ResetThreadPoolThread, set thread names for diagnostics"
  - Aside from that, setting the thread names (at OS level) allows debuggers to identify the threads better as before. For threads that may run user code, the thread Name property is kept as null as before, such that it may be set without exception.
- "Cache-line-separate PortableThreadPool._numRequestedWorkers similarly to coreclr"
  - Was missed before, separated it for consistency
- "Post wait completions to the IO completion port on Windows for coreclr, similarly to before"
  - On Windows, wait completions are queued to the IO thread pool, which is still implemented on the native side. On Unixes, they are queued to the global queue.
- "Reroute managed gate thread into unmanaged side to perform gate activites, don't use unmanaged gate thread"
  - When the config var is enabled, removed the gate thread from the native side. Instead, the gate thread on the managed side calls into the native side to perform gate activities for the IO completion thread pool, and returns a value to indicate whether the gate thread is still necessary.
  - Also added a native-to-managed entry point to request the gate thread to run for the IO completion thread pool
- "Flow config values from CoreCLR to the portable thread pool for compat"
  - Flowed the rest of the thread pool config vars to the managed side, such that COMPlus variables continue to work with the portable thread pool
  - Config var values are stored in AppContext, made the names consistent for supported and unsupported values
- "Port - ..." * 3
  - Ported a few fixes that did not make it into the portable thread pool implementation
- "Fix ETW events"
  - Fixed the EventSource used by the portable thread pool, added missing events
  - For now, the event source uses the same name and GUID as the native side. It seems to work for now for ETW, we may switch to a separate provider (along with updating tools) before enabling the portable thread pool by default.
  - For enqueue/dequeue events, changed to use the object's hash code as the work item identifier instead of the pointer since the pointer may change between enqueue and dequeue
- "Fix perf of counts structs"
  - Structs used for multiple counts with interlocked operations were implemented with explicit struct layout and field offsets. The JIT seems to generate stack-based code for such structs and it was showing up as higher overhead in perf profiles compared to the equivalent native implementation. Slower code in compare-exchange loops can cause a larger gap of time between the read and the compare-exchange, which can also cause higher contention.
  - Changed the structs to use manual bit manipulation instead, and microoptimized some paths. The code is still not as good as that generated by C++, but it seems to perform similarly based on perf profiles.
  - Code size also improved in many cases, for example one of the larger differences was in MaybeAddWorkingWorker(), which decreased from 585 bytes to 382 bytes and with far fewer stack memory operations
- "Fix perf of dispatch loop"
  - Just some minor tweaks as I was looking at perf profiles and code of Dispatch()
- "Fix perf of ThreadInt64PersistentCounter"
  - The implementation used to count completed work items was using `ThreadLocal<T>`, which turned out to be too slow for that purpose according to perf profiles
  - Changed it to rely on the user of the component to provide an object that tracks the count, which the user of the component would obtain from a ThreadStatic field
  - Also removed the thread-local lookup per iteration in one of the hot paths in Dispatch() and improved inlining
- "Miscellaneous perf fixes"
  - A few more small tweaks as I was looking at perf profiles and code
  - In ConcurrentQueue, added check for empty into the fast path
  - For the portable thread pool, updated to trigger the worker thread Wait event after the short spin-wait completes and before actually waiting, the event is otherwise too verbose when profiling and changes performance characteristics
  - Cache-line-separated the gate thread running state as is done in the native implementation
  - Accessing PortableThreadPool.ThreadPoolInstance multiple times was generating less than ideal code that was noticeable in perf profiles. Tried to avoid it especially in hot paths, and in some cases where unnecessary for consistency if nothing else.
  - Removed an extra call to Environment.TickCount in Dispatch() per iteration
  - Noticed that a field that was intended to be cache-line-separated was not actually being separated, see https://github.com/dotnet/runtime/issues/38215, fixed
- "Fix starvation heuristic"
  - Described in comment
- "Implement worker tracking"
  - Implemented the equivalent in the portable thread pool along with raising the relevant event
- "Use smaller stack size for threads that don't run user code"
  - Using the same stack size as in the native side for those threads
- "Note some SOS dependencies, small fixes in hill climbing to make equivalent to coreclr"
  - Corresponds with PR that updates SOS: https://github.com/dotnet/diagnostics/pull/1274
  - Also fixed a couple of things to work similarly to the native implementation
- "Port some tests from CoreRT"
  - Also improved some of the tests
- "Fail-fast in thread pool native entry points specific to thread pool implementations based on config"
  - Scanned all of the managed-to-native entry points from the thread pool and thread-entry functions, and promoted some assertions to be verified in all builds with fail-fast. May help to know in release builds when a path that should not be taken is taken and to avoid running further along that path.
- "Fix SetMinThreads() and SetMaxThreads() to return true only when both changes are successful with synchronization"
  - These are a bit awkward when the portable thread pool is enabled, because they should return true only when both changes are valid and return false without making any changes otherwise, and since the worker thread pool is on the managed side and IO thread pool is on the native side
  - Added some managed-to-native entry points to allow checking validity before making the changes, all under a lock taken by the managed side
- "Fix registered wait removals for fairness since there can be duplicate system wait objects in the wait array"
  - Described in comment
- "Allow multiple DotNETRuntime event providers/sources in EventPipe"
  - Temporary change to EventPipe to be able to get events from dotnet-trace
  - For now, the event source uses the same name and GUID as the native side. It seems to work for now for ETW, and with this change it seems to work with EventPipe for getting events. Subscribing to the NativeRuntimeEventSource does not get thread pool events yet, that is left for later. We may switch to a separate provider (along with updating tools) before enabling the portable thread pool by default, as a long-term solution.
- "Fix registered wait handle timeout logic in the wait thread"
  - The timeout logic was comparing against how long the last wait took and was not timing out waits sometimes, fixed to consider the total time since the last reset of timeout instead
- "Fix Browser build"
  - Updated the Browser-specific thread pool variant based on the other changes

Corresponding PR to update SOS: https://github.com/dotnet/diagnostics/pull/1274
Fixes https://github.com/dotnet/runtime/issues/32020

Commit:
Merge RegisteredWaitHandle implementations

Commit:
Add config var

Commit:
Add dependencies

- Added for coreclr: Unix implementation of LowLevelLifoSemaphore for coreclr, Thread.UninterruptibleSleep0(), Thread.IsThreadPoolThread setter, LowLevelLock
- Moved LowLevelSpinWaiter from mono to shared, restored some functionality from CoreRT

Commit:
Move portable RegisteredWaitHandle implementation to shared ThreadPool.cs

Commit:
Fix timers, tiered compilation, introduced time-sensitive work item queue to simulate coreclr behavior

Commit:
Implement ResetThreadPoolThread, set thread names for diagnostics

Commit:
Cache-line-separate PortableThreadPool._numRequestedWorkers similarly to coreclr

Commit:
Post wait completions to the IO completion port on Windows for coreclr, similarly to before

Commit:
Reroute managed gate thread into unmanaged side to perform gate activites, don't use unmanaged gate thread

Commit:
Flow config values from CoreCLR to the portable thread pool for compat

Commit:
Port - 44970522045f0c323f5735c4fe4b54bd8f71e800 - Fix hill climbing float overflow (dotnet/coreclr#14505)

Commit:
Port - aa5ce2b1bc9ac553ddd493e269a53c999d61e965 - Limit min threads in thread pool to 1 (dotnet/coreclr#14864)

Commit:
Port - 8cc2aa35933677339b9e9ec5485754aa750907df - Optimize AdjustMaxWorkersActive (#1103)

Commit:
Fix ETW events

Commit:
Fix perf of counts structs

Commit:
Fix perf of dispatch loop

Commit:
Fix perf of ThreadInt64PersistentCounter

Commit:
Miscellaneous perf fixes

Commit:
Implement worker tracking

Commit:
Separate portable-only portion of RegisteredWaitHandle

Commit:
Use smaller stack size for threads that don't run user code

Commit:
Note some SOS dependencies, small fixes in hill climbing to make equivalent to coreclr

Commit:
Fix starvation heuristic

Commit:
Port some tests from CoreRT

Commit:
Fail-fast in thread pool native entry points specific to thread pool implementations based on config

Commit:
Fix SetMinThreads() and SetMaxThreads() to return true only when both changes are successful with synchronization

Commit:
Fix registered wait removals for fairness since there can be duplicate system wait objects in the wait array

Commit:
Allow multiple DotNETRuntime event providers/sources in EventPipe

Commit:
Fix registered wait handle timeout logic in the wait thread

Commit:
Fix Browser build

kouvel commented:
Corresponding PR to update SOS: https://github.com/dotnet/runtime/pull/38225

Change is ready for review, but flagging as NO-MERGE for now, until both PRs are ready.

kouvel commented:
Perf data and testing info to follow

kouvel commented:
ASP.NET perf results

.  | Machine | OS | Connections | Clr before | Clr after | Diff | Clr after with PTP | Diff from before | Mono before | Mono after | Diff
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
PlaintextPlatform | 28-proc x64 | Linux | 512 | 14467899 | 14456871 | -0.1% | 14819954 | 2.4% | 9626436 | 9679046 | 0.5%
Plaintext | 28-proc x64 | Linux | 512 | 5265058 | 5286626 | 0.4% | 5344409 | 1.5% | 2855862 | 2890546 | 1.2%
JsonPlatform | 28-proc x64 | Linux | 512 | 1164882 | 1169727 | 0.4% | 1209226 | 3.8% | 830938 | 842005 | 1.3%
Json | 28-proc x64 | Linux | 512 | 954094 | 958050 | 0.4% | 987212 | 3.5% | 619826 | 624145 | 0.7%

No change when portable thread pool is disabled, some small improvements when enabled (some are within the error margin). I wasn't seeing regression or improvement before, but many things changed and hopefully it'll stay.

FortunesPlatform/Fortunes are currently not working. On Windows I was seeing large swings in numbers, without full CPU usage, and resulting in very different numbers from Linux, will try later on a different machine.

marek-safar commented on src/mono/netcore/System.Private.CoreLib/src/System/Threading/ThreadPool.Browser.Mono.cs:
```diff
@@ -10,26 +10,49 @@
+        internal static bool CanSetMinIOCompletionThreads(int ioCompletionThreads) => true;
+        internal static void SetMinIOCompletionThreads(int ioCompletionThreads) { }
+
+        internal static bool CanSetMaxIOCompletionThreads(int ioCompletionThreads) => true;
```
Seems this should be `false`

kouvel replied:
Hmm actually I don't think these methods are needed in this version. It's only called from the portable thread pool implementation to check validity of the value passed in by the user. I'll remove.

marek-safar commented on src/mono/netcore/System.Private.CoreLib/src/System/Threading/ThreadPool.Browser.Mono.cs:
```diff
@@ -10,26 +10,49 @@
+
     public static partial class ThreadPool
     {
+        internal const bool SupportsTimeSensitiveWorkItems = true;
```
Can you add some comment about what this does

kouvel replied:
Done, also changed to false to work similarly to before and avoid creating/checking the extra queue for now

marek-safar commented on src/mono/netcore/System.Private.CoreLib/src/System/Threading/ThreadPool.Browser.Mono.cs:
```diff
@@ -10,26 +10,49 @@
 namespace System.Threading
 {
-    public sealed class RegisteredWaitHandle : MarshalByRefObject
+    public sealed partial class Thread
```
I think this should better be in Tread.Mono.cs

kouvel replied:
Yea it's not the ideal place for it. The build that uses the portable thread pool gets the same implementation from ThreadPool.Portable.cs. I need a coreclr-specific version for now and an "everything else" version, eventually coreclr would also use the common implementation. So perhaps a better place would be in the shared Thread.cs under a `#if !CORECLR` (which can eventually be removed).

marek-safar commented on src/mono/netcore/System.Private.CoreLib/System.Private.CoreLib.csproj:
```diff
@@ -275,7 +275,6 @@
       <Compile Include="$(BclSourcesRoot)\System\Threading\Thread.Mono.cs" />
       <Compile Include="$(BclSourcesRoot)\System\Threading\ThreadPool.Mono.cs" />
       <Compile Include="$(BclSourcesRoot)\System\Threading\LowLevelLock.cs" />
-      <Compile Include="$(BclSourcesRoot)\System\Threading\LowLevelSpinWaiter.cs" />
```
Is `LowLevelLifoSemaphore.Unix.Mono.cs` still needed?

kouvel replied:
Yea coreclr and mono currently have different ways of getting the LIFO wait release order. Coreclr's version creates a regular semaphore and the LifoSemaphore's wait issues a prioritized wait in the PAL such that the waiter is registered in stack order instead of the usual queue order, release works the same way as a dequeue, so prioritized waits would be released in LIFO order. If the waiter registration order could be similarly tweaked in mono then we can have a simpler runtime call to do that and share most of the implementations. On Windows coreclr uses IO completion port so it would continue have a different implementation but that may be shareable too for Windows.

marek-safar replied:
My preference would be to have same implementation. @lambdageek is this something we can easily tweak?

lambdageek replied:
@marek-safar if I follow the question is whether we can share most of the managed implementation which uses a normal managed semaphore and has a separate QCall that queues the waiting thread ahead of the others?  If yes, I think that will be a bit challenging for us.  At the lowest level we implement waiting on a semaphore by waiting on a pthread condition variable. pthread condition variables don't give us a way to affect the queue order (as far as I know).

kouvel replied:
I see, how does `WaitHandle.WaitAll` with a bunch of semaphores work in mono?

It seems possible that mono's implementation may be simpler and more performant than coreclr's, just trying to understand differences and reasons for them if any.

lambdageek replied:
`WaitAll` is done by waiting on a global condvar. 

https://github.com/dotnet/runtime/blob/bc782a44d942c58799678f6bbac716a298928bce/src/mono/mono/metadata/w32handle.c#L40-L46

kouvel replied:
I see, so when any wait object that is a component of a multi-wait is signaled, all threads waiting on it would be released and each thread would, for a wait-all, I imagine, acquire the mutex for each of the wait objects involved in deterministic order to determine if all of the waits are satisfied?

That might be an advantage in coreclr's PAL implementation (which is similar to Windows) where the signaling thread determines whether multi-waits of a waiter are satisfied before releasing it, thereby releasing a minimum number of threads.

kouvel commented:
Microbenchmark perf. The benchmark measures throughput with a very short CPU-intensive delay in each work item, trying to measure mostly the overhead of the thread pool for very short work items, in sustained fashion or in bursty fashion. The length of the burst is the number multiplied by proc count. The benchmark is mostly useful to find larger regressions, sometimes even large differences don't translate into reality, especially as work items do more work.

Windows x64 8-proc

.  | Clr before | Clr after | Diff | Clr after with PTP | Diff from before
--: | --: | --: | --: | --: | --:
Global sustained | 9334 | 9301 | -0.3% | 9313 | -0.2%
Global 1*proc burst | 1181 | 1183 | 0.2% | 1175 | -0.5%
Global 4*proc burst | 3363 | 3420 | 1.7% | 3350 | -0.4%
Global 16*proc burst | 5809 | 5898 | 1.5% | 5859 | 0.9%
Global 64*proc burst | 6973 | 7019 | 0.7% | 6978 | 0.1%
Global 256*proc burst | 7148 | 7264 | 1.6% | 7191 | 0.6%
Local sustained | 19968 | 20711 | 3.7% | 21423 | 7.3%
Local 1*proc burst | 1174 | 1177 | 0.2% | 1168 | -0.5%
Local 4*proc burst | 3684 | 3725 | 1.1% | 3767 | 2.2%
Local 16*proc burst | 8705 | 8818 | 1.3% | 9036 | 3.8%
Local 64*proc burst | 13433 | 13605 | 1.3% | 14184 | 5.6%
Local 256*proc burst | 15435 | 15560 | 0.8% | 16188 | 4.9%

Linux x64 8-proc VM

.  | Clr before | Clr after | Diff | Clr after with PTP | Diff from before | Mono before | Mono after | Diff | Notes for Mono
--: | --: | --: | --: | --: | --: | --: | --: | --: | --:
Global sustained | 8954 | 8827 | -1.4% | 8747 | -2.3% | 7098 | 7508 | 5.8% |  
Global 1*proc burst | 121 | 114 | -5.5% | 132 | 9.4% | 35 | 41 | 18.7% |  
Global 4*proc burst | 1670 | 1696 | 1.5% | 1694 | 1.4% | 104 | 122 | 17.1% |  
Global 16*proc burst | 4125 | 4111 | -0.3% | 4230 | 2.5% | 3056 | 3195 | 4.6% |  
Global 64*proc burst | 6005 | 6025 | 0.3% | 6017 | 0.2% | 5054 | 5256 | 4.0% |  
Global 256*proc burst | 6719 | 6695 | -0.4% | 6765 | 0.7% | 5846 | 6068 | 3.8% |  
Local sustained | 16499 | 17801 | 7.9% | 19305 | 17.0% | 9328 | 10261 | 10.0% |  
Local 1*proc burst | 114 | 110 | -4.1% | 125 | 9.5% | 34 | 46 | 36.5% | High error
Local 4*proc burst | 1634 | 1698 | 3.9% | 1805 | 10.4% | 153 | 235 | 53.3% | Very high error
Local 16*proc burst | 5219 | 5275 | 1.1% | 5518 | 5.7% | 2170 | 3485 | 60.7% | Very high error
Local 64*proc burst | 10008 | 10283 | 2.7% | 10845 | 8.4% | 6193 | 6573 | 6.1% |  
Local 256*proc burst | 12883 | 13395 | 4.0% | 14335 | 11.3% | 7708 | 8362 | 8.5% |  

For Clr results, the tests with the regressions appear to be multimodal, not sure why. I don't think it's significant.

Those three tests when running under Mono seem to have very high error margins before and after the change, they can be ignored. I had collected the Mono perf numbers before, when my machine was reporting lower numbers on all runtimes, my machine does that sometimes.

Windows arm64 8-proc

. | Clr before | Clr after | Diff | Clr after with PTP | Diff from before
--: | --: | --: | --: | --: | --:
Global sustained | 4600 | 4611 | 0.2% | 4547 | -1.2%
Global 1*proc burst | 562 | 575 | 2.3% | 645 | 14.8%
Global 4*proc burst | 1458 | 1475 | 1.2% | 1645 | 12.9%
Global 16*proc burst | 2684 | 2725 | 1.5% | 2813 | 4.8%
Global 64*proc burst | 3298 | 3340 | 1.3% | 3343 | 1.4%
Global 256*proc burst | 3468 | 3506 | 1.1% | 3489 | 0.6%
Local sustained | 8662 | 8667 | 0.1% | 8858 | 2.3%
Local 1*proc burst | 557 | 574 | 2.9% | 644 | 15.7%
Local 4*proc burst | 1540 | 1545 | 0.3% | 1792 | 16.4%
Local 16*proc burst | 3464 | 3565 | 2.9% | 3748 | 8.2%
Local 64*proc burst | 5102 | 5271 | 3.3% | 5234 | 2.6%
Local 256*proc burst | 5371 | 5412 | 0.8% | 5509 | 2.6%

kouvel commented:
General testing done or pending:
- ThreadPool, Timer, and Thread tests
- Thread pool features - starvation on worker threads, starvation on IO completion threads, hill climbing
- Anything that was added like each reroute based on config, worker tracking, etc.
- SOS `ThreadPool -ti -wi`
- PerfView, perfcollect, dotnet-trace (testing on mono pending), dotnet-monitor (pending)
  - perfcollect was not showing EventSource events including existing ones, so when the portable thread pool is enabled those would not show up currently
- Some remaining sanity testing (pending)

kouvel commented:
Looks like Mono folks are already added, CC some more people
- For general changes: @janvorli @jkoritzinsky @jkotas @mangod9 @stephentoub 
- For events: @brianrob @josalem @noahfalk @sywhang

mangod9 commented on src/libraries/System.Private.CoreLib/src/System/Threading/LowLevelLock.cs:
```diff
@@ -0,0 +1,217 @@
+
+        public void VerifyIsLocked()
+        {
+#if DEBUG
```
should the func defs all be under `#if DEBUG` since they have debug only code? Dont believe the compiler would optimize them out on non-debug?

kouvel replied:
I wanted to attribute them with `[Conditional("DEBUG")]` but those methods are still compiled in release, so using debug-only variables like `_ownerThread` was yielding a compile error. I think they can still be attributed at such, just need an `#if DEBUG` around debug-only variable uses. Just wanted to avoid the nuisance of `#if'ing` out all calls to them. Even without the attribute, these would be empty methods in release build, the JIT should be able to optimize out those calls. In any case the attribute should get the best of both, as it looks like those methods don't even get IL generated for them in release builds.

jkotas replied:
If you attribute them with `[Conditional("DEBUG")]`, there won't be any calls to them and the IL linker will strip the methods in the release binary.

mangod9 commented on src/libraries/System.Private.CoreLib/src/System/Threading/ThreadInt64PersistentCounter.cs:
```diff
@@ -2,79 +2,69 @@
-            Debug.Assert(_threadLocalNode.Value == null);
+            var node = new ThreadLocalNode(this);
 
+            s_lock.Acquire();
```
any reason not to acquire the lock within try/finally?

kouvel replied:
Since thread-abort is not supported it doesn't make much difference. Only thing is if Acquire were to throw (this particular method currently doesn't allocate) then the call to Release in the finally block would be invalid. In order to weed out those issues when calling it inside the try block it would have to return something or have an `out acquired` parameter (similar solution for handling thread-abort). It feels simpler to call it outside the try, since Acquire would already have a contract where if it throws, then it must not have acquired the lock, though even that may be difficult to guarantee for all types of exceptions.

janvorli replied:
Thread abort is actually supported internally - debugger funceval support uses it to abort long running or hung methods or properties.

jkotas replied:
We do not make guarantees about the state of the process when debugger thread abort kicks in.

kouvel replied:
Based on https://github.com/dotnet/coreclr/pull/8551 and https://github.com/dotnet/coreclr/pull/8949 it looks like the JIT would remove try/finally blocks where the try block is empty and the finally block has code. So I figure there would not be a way to harden for thread-abort during func-eval, unless there is some other way.

brianrob commented on src/libraries/System.Private.CoreLib/src/System/Threading/PortableThreadPoolEventSource.cs:
```diff
@@ -3,117 +3,286 @@
 {
-    [EventSource(Name = "Microsoft-Windows-DotNETRuntime", Guid = "{e13c0d23-ccbc-4e12-931b-d9cc2eee27e4}")]
-    public sealed class PortableThreadPoolEventSource : EventSource
+    [EventSource(Name = "Microsoft-Windows-DotNETRuntime", Guid = "e13c0d23-ccbc-4e12-931b-d9cc2eee27e4")]
```
Per our conversation offline, you should name this something else, like "Microsoft-DotNETRuntime-PortableThreadPool".  You should also remove the GUID so that EventSource can compute the GUID from the specified name.  This will put you on the supported path, until there is opportunity for the diagnostics folks to work through the details of potentially using NativeRuntimeEventSource.

jkotas replied:
> You should also remove the GUID so that EventSource can compute the GUID from the specified name

We have been using the explicit Guid on all event sources in CoreLib to avoid the expensive reflection to compute it. I do not see a good reason for this one to be on a different plan.

brianrob replied:
Yes, you're right.  @stephentoub and @benaadams mentioned this below.  I had forgotten that this pattern is being used inside the framework.  The GUID just needs to match what EventSource would have generated so that tools don't need to know the GUID, just the name.

kouvel replied:
From our further discussion I have changed the name/guid in mono and undid the EventPipe change in mono, and kept the same name/guid for coreclr and conditioned the EventPipe change for when the portable thread pool is enabled (no change by default). At the moment that seems to be the easiest way to get a similar experience to before in perf profiles.

brianrob commented on src/libraries/System.Private.CoreLib/src/System/Threading/PortableThreadPoolEventSource.cs:
```diff
@@ -3,117 +3,286 @@
 
         private PortableThreadPoolEventSource()
+            : base(
+                  new Guid(0xe13c0d23, 0xccbc, 0x4e12, 0x93, 0x1b, 0xd9, 0xcc, 0x2e, 0xee, 0x27, 0xe4),
```
You should remove these parameters to the base class constructor and just let EventSource pick these up from the attribute.

stephentoub replied:
Other PRs have explicitly switched to using this base ctor to improve startup time / reduce reflection usage on startup, e.g.
https://github.com/dotnet/coreclr/pull/21714

benaadams replied:
Not using the explicit Guid is harsh for startup since so much code needs to be Jitted

brianrob replied:
Ah, yes that seems fine.  Just make sure that the GUID matches what EventSource would have generated for the name, so that tools still function properly if you don't specify the GUID, but instead specify the name.

benaadams replied:
Wonder if there is a build verification step that can be added that checks the GUIDs against ones that would be generated?

brianrob replied:
Definitely possible, but I'm not too concerned about this, just because it's essentially limited to EventSources in this repo.  If we wanted to find a way to fix this up more publicly, then I think we'd need to try and fix the API, or if we couldn't find a good way, then yes, a tool to validate.

brianrob commented on src/libraries/System.Private.CoreLib/src/System/Threading/PortableThreadPoolEventSource.cs:
```diff
@@ -3,117 +3,286 @@
-        [Event(2, Level = EventLevel.Informational, Message = WorkerThreadMessage, Task = Tasks.WorkerThreadTask, Opcode = EventOpcode.Stop, Version = 0, Keywords = Keywords.ThreadingKeyword)]
-        public void WorkerThreadStop(short numExistingThreads)
+        [Event(50, Level = EventLevel.Informational, Message = Messages.WorkerThread, Task = Tasks.ThreadPoolWorkerThread, Opcode = EventOpcode.Start, Version = 0, Keywords = Keywords.ThreadingKeyword)]
+        public unsafe void ThreadPoolWorkerThreadStart(int numExistingThreads)
```
Most of these events, as implemented, aren't going to work.  They will be emitted properly, but won't really be consumable in a supported way.  The reason is that the method signature needs to match the event arguments.  When the EventSource is instantiated, it creates a manifest by walking the set of event methods, and building up their metadata through reflection over the method arguments.  Then, the manifest is emitted into the event stream so that tools can consume it and have knowledge of how to decode the events.  As an example, in the case of ThreadPoolWorkerThreadStart, the EventSource generated manifest will state that the only payload argument is an int called numExistingThreads.  In reality, you're actually writing out 3 payload arguments.

I would recommend simplifying this code a bit, and using the standard EventSource pattern of:

```
[Event(id, ...)]
public void EventName(payloadArgs...)
{
    WriteEvent(id, payloadArgs...);
}
```

This should be all that you need.  You also shouldn't need to call IsEnabled inside of the event method - this will happen inside of WriteEvent, and at this point, you've already computed the arguments.  If you do need to compute expensive arguments, you can use a [NonEvent] method as an outer wrapper around an event method to make sure that you only compute the arguments if the event is enabled.

Totally fine to keep all of the existing Keyword, OpCode, Task and Version stuff, though Version=0 is the default.

kouvel replied:
Thanks, fixed. Due to the ClrInstanceId being ushort, I couldn't use any existing overload of WriteEvent().

brianrob commented on src/libraries/System.Private.CoreLib/src/System/Threading/PortableThreadPoolEventSource.cs:
```diff
@@ -3,117 +3,286 @@
+        // This value does not seem to be used, leaving it as zero for now. It may be useful for a scenario that may involve
+        // multiple instances of the runtime within the same process, but then it seems unlikely that both instances' thread
+        // pools would be in moderate use.
+        private const short ClrInstanceId = 0;
```
Yup, this is a historical artifact.  Seems fine to leave it if you're trying to match the exact arguments as the native versions of these events.  Otherwise, you could get rid of this argument.

brianrob commented on src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/FrameworkEventSource.cs:
```diff
@@ -94,25 +95,27 @@ public void ThreadPoolEnqueueWork(long workID)
-            // convert the Object Id to a long
-            ThreadPoolEnqueueWork((long)*((void**)Unsafe.AsPointer(ref workID)));
-        }
+        [MethodImpl(MethodImplOptions.NoInlining)]
```
Why do these methods need to be marked as non-inlineable?

kouvel replied:
These are called from hot paths and inlining them makes the caller's code larger and the hot path more sparse

brianrob replied:
Gotcha.  Makes sense.

brianrob commented:
>         // id - is a managed object. it gets translated to the object's address. ETW listeners must

Since you've switched to GetHashCode for the identifier, then you'll want to also remove the comments that talk about the requirement of consumers to keep track of object movement.

---
Refers to: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/FrameworkEventSource.cs:134 in 247604c. [](commit_id = 247604cdf29cfabdde91e0c578b665e99ad7747c, deletion_comment = False)

Commit:
Remove unnecessary methods from Browser thread pool, address some feedback

kouvel commented:
CscRoslynSource steady-state perf - Roslyn compiling itself repeatedly in the same process. No change from before.

Windows x64 8-proc

Clr   before | Clr after | Diff | Clr after with PTP | Diff from before
--: | --: | --: | --: | --:
599.6 | 600.3 | 0.1% | 600.1 | 0.1%

kouvel commented:
Rebased to fix conflicts

Commit:
Address feedback for events, undo EventPipe change in mono, change event source name in mono, add comments

Commit:
Fix build warning after merge

kouvel commented:
Rebased to fix conflicts

Commit:
Disable new timer test for browser

kouvel commented on src/coreclr/src/vm/eventpipeconfiguration.cpp:
```diff
@@ -147,10 +147,16 @@ bool EventPipeConfiguration::RegisterProvider(EventPipeProvider &provider, Event
+    // different sets of events by component through different event sources.
+    // TODO: This change to allow multiple DotNETRuntime providers is temporary to get EventPipe working for
+    // PortableThreadPoolEventSource. Once a long-term solution is figured out, this change should be reverted.
+    if (!provider.GetProviderName().Equals(W("Microsoft-Windows-DotNETRuntime")))
```
Yep makes sense. Went with option (b) for now.

Commit:
Fix EventSource tests to expect the new provider name used in mono

noahfalk commented on src/coreclr/src/inc/clrconfigvalues.h:
```diff
@@ -554,10 +554,12 @@ RETAIL_CONFIG_DWORD_INFO(INTERNAL_ThreadSuspendInjection, W("INTERNAL_ThreadSusp
 ///
 /// Threadpool
 ///
+RETAIL_CONFIG_DWORD_INFO(INTERNAL_ThreadPool_UsePortableThreadPool, W("ThreadPool_UsePortableThreadPool"), 0, "Uses the managed portable thread pool implementation instead of the unmanaged one.")
```
Now that this config var is also controlling alternate event pipe behavior I want to make sure that nobody changes the default value of the switch until the EventPipe issue is resolved permanently. Given that the name of the switch doesn't make this connection obvious I'd suggest doing two things:
1. Create an issue tracking that this work needs to happen before the switch value can be changed.
2. Add a comment calling out the non-obvious dependency and including a link to the issue.

kouvel replied:
Filed https://github.com/dotnet/runtime/issues/38763 and added links to it in a few places

jkotas commented on src/coreclr/src/System.Private.CoreLib/src/System/Threading/Thread.CoreCLR.cs:
```diff
@@ -333,7 +341,14 @@ private void SetStartHelper(Delegate start, int maxStackSize)
+                SetBackgroundNative(value);
+                if (!value && !_mayNeedResetForThreadPool)
+                {
+                    _mayNeedResetForThreadPool = value;
```
This looks like a no-op. Both `value` and `_mayNeedResetForThreadPool` will be fals when this branch is taken.

Should this be `_mayNeedResetForThreadPool = true` ?

Also, do we need a test to catch this?

kouvel replied:
Nice catch, yea my intention was `= true`. Fixed and updated the test to verify each state reset separately.

jkotas commented on src/coreclr/src/System.Private.CoreLib/src/System/Threading/Thread.CoreCLR.cs:
```diff
@@ -333,7 +341,14 @@ private void SetStartHelper(Delegate start, int maxStackSize)
+            set
+            {
+                SetBackgroundNative(value);
+                if (!value && !_mayNeedResetForThreadPool)
```
```suggestion
                if (!value)
```

I would not worry about trying to avoid the extra write.

jkotas commented on src/coreclr/src/System.Private.CoreLib/src/System/Threading/Thread.CoreCLR.cs:
```diff
@@ -347,13 +362,22 @@ public bool IsBackground
+            set
+            {
+                SetPriorityNative((int)value);
+                if (value != ThreadPriority.Normal && !_mayNeedResetForThreadPool)
```
```suggestion
                if (value != ThreadPriority.Normal)
```

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Threading/ThreadInt64PersistentCounter.cs:
```diff
@@ -2,79 +2,69 @@
-        private readonly ThreadLocal<ThreadLocalNode> _threadLocalNode = new ThreadLocal<ThreadLocalNode>(trackAllValues: true);
         private long _overflowCount;
 
+        // This should be a HashSet, that type is currently not available in CoreLib
```
This is not the case anymore. HashSet was moved to CoreLib.

kouvel replied:
Updated to use HashSet

Commit:
Revert event source name in mono to match coreclr

Commit:
Add issue link for prerequisites of enabling the portable thread pool by default

vargaz commented on src/mono/mono/metadata/threads.c:
```diff
@@ -2022,8 +2022,13 @@ ves_icall_System_Threading_Thread_SetName_icall (MonoInternalThreadHandle thread
+	// appropriate.
 	mono_thread_set_name (mono_internal_thread_handle_ptr (thread_handle),
-		name8, (gint32)name8_length, name16, MonoSetThreadNameFlag_Permanent, error);
+		name8, (gint32)name8_length, name16, MonoSetThreadNameFlag_Reset, error);
```
This change should probably be inside an #ifdef ENABLE_NETCORE.

kouvel replied:
Thanks, fixed

Commit:
Address feedback

kouvel commented:
Regarding commit titled "Revert event source name in mono to match coreclr", after discussion with @lateralusX we had decided to keep the same event source name/guid in mono as in coreclr for now at least for consistency.

fanyang-mono commented:
> ASP.NET RPS perf results
> 
> .	Machine	OS	Connections	Clr before	Clr after	Diff	Clr after with PTP	Diff from before	Mono before	Mono after	Diff
> PlaintextPlatform	28-proc x64	Linux	512	14467899	14456871	-0.1%	14819954	2.4%	9626436	9679046	0.5%
> Plaintext	28-proc x64	Linux	512	5265058	5286626	0.4%	5344409	1.5%	2855862	2890546	1.2%
> JsonPlatform	28-proc x64	Linux	512	1164882	1169727	0.4%	1209226	3.8%	830938	842005	1.3%
> Json	28-proc x64	Linux	512	954094	958050	0.4%	987212	3.5%	619826	624145	0.7%
> No change when portable thread pool is disabled, some small improvements when enabled (some are within the error margin). I wasn't seeing regression or improvement before, but many things changed and hopefully it'll stay.
> 
> FortunesPlatform/Fortunes are currently not working. On Windows I was seeing large swings in numbers, without full CPU usage, and resulting in very different numbers from Linux, will try later on a different machine. I wasn't able to test on arm64, as updating the build with my locally cross-built native binaries doesn't seem to work (even without any changes), I can verify after it's merged and an sdk with the change is produced.

Hi Kount, I am trying to understand the Mono results here. I wonder when you mention Mono, did you mean Mono with JIT? Mono has multiple flavors, such as JIT, LLVM JIT, AOT, LLVM AOT and so on. LLVM AOT is the flavor that the mobile customers use. I wonder if that should be tested as well for this PR.

lambdageek commented on src/libraries/Native/Unix/System.Native/pal_threading.h:
```diff
@@ -0,0 +1,133 @@
+} LowLevelMutex;
+
+bool LowLevelMutex_TryInitialize(LowLevelMutex *mutex);
+bool LowLevelMutex_TryInitialize(LowLevelMutex *mutex)
```
C does not have the One Definition Rule.  All these functions in the header should be `static inline`

kouvel replied:
I see, thanks that worked

vargaz commented:
The mono changes look ok to me.

kouvel commented:
Rebased to fix conflicts

kouvel commented:
Has anyone had a chance to look at changes to the portable thread pool and other shared portions of the libraries? Also the coreclr native portions? Might also be easier to review by commit.

kouvel commented:
Issue for failure in "Libraries Test Run release mono OSX x64 Debug": https://github.com/dotnet/runtime/issues/39109

Commit:
Fix race condition in registered wait unregister

- For a registered wait that is automatically unregistered (due to `executeOnlyOnce: true`), the registered wait handle gets added to the array of pending removals, and this automatic unregister does not wait for the removal to actually happen
- If shortly after that a user calls `Unregister(null)` on the same registered wait handle, it is supposed to wait for the removal to actually happen, but was not because the handle is already in the array of pending removals
- A `Dispose` on the wait handle shortly after `Unregister` returns would delete the safe handle and `DangerousRelease` upon removal would throw and crash the process
- Fixed by waiting when a registered wait handle is pending removal, regardless of whether the caller of `Unregister` added the handle to the array of pending removals or if it was added by anyone else

kouvel commented:
It doesn't look like this will make it into .NET 5, pushed to .NET 6

Commit:
Fix a new issue in WaitThread after shifting items in arrays

kouvel commented:
Rebased to fix conflicts

kouvel commented:
Rebased to fix conflicts. Separated the originally second commit into 3 separate commits to make it easier to review, so the originally first three commits are now the first five commits, commit descriptions have been adjusted inline above.

Commit:
Fix usage of LowLevelMonitor after rebase

marek-safar commented on src/mono/netcore/System.Private.CoreLib/src/System/Threading/ThreadPool.Browser.Mono.cs:
```diff
@@ -4,34 +4,40 @@
 
 namespace System.Threading
 {
-    [UnsupportedOSPlatform("browser")]
```
This needs to come back to align with ref version

kouvel replied:
I had changed `RegisteredWaitHandle` to be a partial class and kept the attribute in the shared implementation in `ThreadPool.cs`

marek-safar commented on src/mono/netcore/System.Private.CoreLib/src/System/Threading/ThreadPool.Browser.Mono.cs:
```diff
@@ -76,36 +82,20 @@ internal static void RequestWorkerThread()
-                throw new ArgumentNullException(nameof(callBack));
+        internal static object? GetOrCreateThreadLocalCompletionCountObject() => null;
 
+        private static void RegisterWaitForSingleObjectCore(WaitHandle? waitObject, RegisteredWaitHandle registeredWaitHandle) =>
```
Which public API does call this?

kouvel replied:
This is called by variants of `ThreadPool.RegisterWaitForSingleObject`. The null checks and some common code were moved to the shared `ThreadPool.cs` in the commit titled "Merge RegisteredWaitHandle implementations".

marek-safar commented:
@jkotas could you help with this PR to be merged quickly? It changes the default implementation for Mono which I'd rather start testing in wild asap.

janvorli commented:
@marek-safar I've started reviewing this change with @kouvel two weeks ago (together with few other folks) to get confidence about all the changes. We plan to have two more meeting next week and then I hope I'll be able to approve this PR.

Commit:
Fix tests that use RemoteExecutor to include a test for whether it is supported, fix assert for browser builds

Commit:
Fix build

SamMonoRT commented:
fyi @fanyang-mono to track perf improvements for Mono.

Commit:
Slight fix to starvation heuristic to be closer to what it was before

janvorli commented on src/libraries/System.Threading.ThreadPool/tests/RegisteredWaitTests.cs:
```diff
@@ -0,0 +1,518 @@
+        }
+
+        [ConditionalFact(typeof(PlatformDetection), nameof(PlatformDetection.IsThreadingSupported))]
+        public static void StateIsPasssedThroughToCallback()
```
A nit:
```suggestion
        public static void StateIsPassedThroughToCallback()
```

janvorli commented on src/coreclr/src/vm/comthreadpool.cpp:
```diff
@@ -446,6 +555,32 @@ FCIMPL5(LPVOID, ThreadPoolNative::CorRegisterWaitForSingleObject,
+    _ASSERTE_ALL_BUILDS(__FILE__, ThreadpoolMgr::UsePortableThreadPool());
+
+    HANDLE completeWaitWorkItemHandle = NULL;
+    struct _gc
```
A nit - since you need to protect just a single object, you can use the HELPER_METHOD_FRAME_BEGIN_1 version without creating the _gc struct. There are also variants for 2 and 3 objects, only when using more objects you need to use the HELPER_METHOD_FRAME_BEGIN_PROTECT.
But please feel free to keep it as is if you prefer it that way.

kouvel commented:
Rebased to latest

Commit:
Fix license headers in new files

kouvel commented on src/libraries/System.Private.CoreLib/src/System/Threading/RegisteredWaitHandle.Portable.cs:
```diff
@@ -0,0 +1,84 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
```
Looks like it was removed in other files, I have now removed it from new files

Commit:
Address feedback

kouvel commented on src/coreclr/src/vm/comthreadpool.cpp:
```diff
@@ -122,6 +122,107 @@ DelegateInfo *DelegateInfo::MakeDelegateInfo(OBJECTREF *state,
+            *appContextConfigNameRef = NULL;
+            return nextConfigVariableIndex + 1;
+
+        case 1: if (TryGetConfig(CLRConfig::INTERNAL_ThreadPool_ForceMinWorkerThreads, false, W("System.Threading.ThreadPool.MinThreads"))) { return nextConfigVariableIndex; } // fall through
```
Removed `nextConfigVariableIndex` and hardcoded the return values (always case label + 1)

kouvel commented on src/libraries/System.Private.CoreLib/src/System/Threading/LowLevelLock.cs:
```diff
@@ -0,0 +1,217 @@
+    /// </summary>
+    internal sealed class LowLevelLock : IDisposable
+    {
+        // TODO: Tune these values
```
They haven't needed tuning so far, I don't think these locks are hot enough yet. We can decide later whether to tune, removed the comment for now.

kouvel commented on src/libraries/System.Private.CoreLib/src/System/Threading/LowLevelSpinWaiter.cs:
```diff
@@ -5,8 +5,52 @@
+                // resource being waited upon
+                if (spinningThreadCount <= processorCount)
+                {
+                    // For uniprocessor systems, start at the yield threshold since the pause instructions used for waiting
```
The benefit of spin-waiting would also be lower on containers affinitized to 1 processor, although the yield may allow other processes to do work on the core, if the spin-waiting is waiting for a change that would be made by a different thread in the same process (often the case) then it would not be able to see the change until it allows a different thread to run. `Sleep` may be slightly cheaper than the actual wait sometimes (used in `LowLevelLock` in this case, where the actual wait involves acquiring a mutex and waiting on a condition variable), but it may be slightly worse in other cases where there is no other thread to schedule (unlikely in this case). In several other cases spin-waiting is avoided on single-processor machines. The spin-waiting done here is minimal and probably decent for now.

kouvel commented on src/libraries/System.Private.CoreLib/src/System/Threading/PortableThreadPool.cs:
```diff
@@ -11,11 +12,8 @@ namespace System.Threading
-#pragma warning restore IDE1006 // Naming Styles
-
         private const int ThreadPoolThreadTimeoutMs = 20 * 1000; // If you change this make sure to change the timeout times in the tests.
+        private const int SmallStackSizeBytes = 256 * 1024;
```
I think the stack size of thread pool threads that run user code should also be configurable (currently they always use the default stack size). It seems like a lot of threads would be necessary to exhaust the address space on 32-bit platforms, the machine would likely run out of physical memory before that. It makes sense to use a smaller stack size by default on 32-bit platforms, perhaps for all threads. In this case the small stack size is used for thread pool threads that don't run user code, it would have to be tested to see if lowering the size would be ok. For now I'm just using the same stack size as the native side and it seems to be sufficient from my testing.

marek-safar commented on src/libraries/System.Private.CoreLib/src/System/Threading/PortableThreadPool.HillClimbing.cs:
```diff
@@ -12,32 +12,35 @@ internal partial class PortableThreadPool
-                    waveMagnitudeMultiplier: AppContextConfigHelper.GetInt32Config("HillClimbing_WaveMagnitudeMultiplier", 100, false) / 100.0,
-                    waveHistorySize: AppContextConfigHelper.GetInt32Config("HillClimbing_WaveHistorySize", 8, false),
-                    targetThroughputRatio: AppContextConfigHelper.GetInt32Config("HillClimbing_Bias", 15, false) / 100.0,
-                    targetSignalToNoiseRatio: AppContextConfigHelper.GetInt32Config("HillClimbing_TargetSignalToNoiseRatio", 300, false) / 100.0,
-                    maxChangePerSecond: AppContextConfigHelper.GetInt32Config("HillClimbing_MaxChangePerSecond", 4, false),
-                    maxChangePerSample: AppContextConfigHelper.GetInt32Config("HillClimbing_MaxChangePerSample", 20, false),
-                    sampleIntervalMsLow: AppContextConfigHelper.GetInt32Config("HillClimbing_SampleIntervalLow", DefaultSampleIntervalMsLow, false),
-                    sampleIntervalMsHigh: AppContextConfigHelper.GetInt32Config("HillClimbing_SampleIntervalHigh", DefaultSampleIntervalMsHigh, false),
-                    errorSmoothingFactor: AppContextConfigHelper.GetInt32Config("HillClimbing_ErrorSmoothingFactor", 1, false) / 100.0,
-                    gainExponent: AppContextConfigHelper.GetInt32Config("HillClimbing_GainExponent", 200, false) / 100.0,
-                    maxSampleError: AppContextConfigHelper.GetInt32Config("HillClimbing_MaxSampleErrorPercent", 15, false) / 100.0
+                return new HillClimbing(wavePeriod: AppContextConfigHelper.GetInt32Config("System.Threading.ThreadPool.HillClimbing.WavePeriod", 4, false),
+                    maxWaveMagnitude: AppContextConfigHelper.GetInt32Config("System.Threading.ThreadPool.HillClimbing.MaxWaveMagnitude", 20, false),
+                    waveMagnitudeMultiplier: AppContextConfigHelper.GetInt32Config("System.Threading.ThreadPool.HillClimbing.WaveMagnitudeMultiplier", 100, false) / 100.0,
+                    waveHistorySize: AppContextConfigHelper.GetInt32Config("System.Threading.ThreadPool.HillClimbing.WaveHistorySize", 8, false),
+                    targetThroughputRatio: AppContextConfigHelper.GetInt32Config("System.Threading.ThreadPool.HillClimbing.Bias", 15, false) / 100.0,
+                    targetSignalToNoiseRatio: AppContextConfigHelper.GetInt32Config("System.Threading.ThreadPool.HillClimbing.TargetSignalToNoiseRatio", 300, false) / 100.0,
+                    maxChangePerSecond: AppContextConfigHelper.GetInt32Config("System.Threading.ThreadPool.HillClimbing.MaxChangePerSecond", 4, false),
+                    maxChangePerSample: AppContextConfigHelper.GetInt32Config("System.Threading.ThreadPool.HillClimbing.MaxChangePerSample", 20, false),
+                    sampleIntervalMsLow: AppContextConfigHelper.GetInt32Config("System.Threading.ThreadPool.HillClimbing.SampleIntervalLow", DefaultSampleIntervalMsLow, false),
```
Could this be done via static properties instead of eager reading into constructor? It'd allow us to optimize this away for configurations without AppContext easier

kouvel replied:
Would it work to read the config values in the constructor itself (similarly through `AppContextConfigHelper`) instead of passing them as arguments to the constructor?

marek-safar replied:
It should help

kouvel replied:
Ok I have done that now, it's a bit cleaner too

Commit:
Change pattern for accessing the event source for better ILLinker compatibility

kouvel commented on src/libraries/System.Private.CoreLib/src/System/Threading/PortableThreadPool.GateThread.cs:
```diff
@@ -13,114 +16,146 @@ private static class GateThread
-                _ = s_cpu.CurrentUtilization; // The first reading is over a time range other than what we are focusing on, so we do not use the read.
+                bool disableStarvationDetection =
+                    AppContextConfigHelper.GetBooleanConfig("System.Threading.ThreadPool.DisableStarvationDetection", false);
+                bool debuggerBreakOnWorkStarvation =
```
Are there cases where `Debugger.Break` is not supported, and what is the way to check for that?

marek-safar replied:
Hmm, there is no general property for that. I guess we can pass on this for now

Commit:
Move hill climbing config reads into constructor instead of passing them as parameters

kouvel commented on src/libraries/System.Private.CoreLib/src/System/Threading/PortableThreadPool.WaitThread.cs:
```diff
@@ -20,23 +21,26 @@ internal partial class PortableThreadPool
         internal void RegisterWaitHandle(RegisteredWaitHandle handle)
         {
+            PortableThreadPoolEventSource log = PortableThreadPoolEventSource.Log;
+            if (log.IsEnabled())
```
Changed to use `PortableThreadPoolEventSource.Log` instead of loading the log into a variable. I figured the load of the log would be unused and removed when `IsEnabled()` is statically false, not sure.

kouvel commented:
Rebased to fix conflicts, last two commits are new

kouvel commented:
Rebased to fix conflict
